app/util/poll.js

```diff
@@ -3,17 +3,47 @@ import delay from './delay'
 const DEFAULT_ATTEMPTS = Infinity
 const DEFAULT_FREQUENCY = 1000
 
-export default function poll(
+export class CancellationError extends Error {
+  constructor() {
+    super('Polling was actively cancelled')
+    this.name = 'CancellationError'
+  }
+}
+
+export class TimeoutError extends Error {
+  constructor() {
+    super('Polling reached max attempts')
+    this.name = 'TimeoutError'
+  }
+}
+
+export function cancellablePoll(
   request,
   { attempts = DEFAULT_ATTEMPTS, frequency = DEFAULT_FREQUENCY } = {}
 ) {
-  return request().catch(function retry(err) {
+  let hasCancelled = false
+  const promise = request().catch(function retry() {
```

Comment:
Maybe Im missing something but wouldn't it be better to define `retry` outside of this catch block so that it doesn't get redefined every time `request` throws an error?

Reply:
Indeed. I'll change, test and PR separately.

Reply:
Pull request #1734